service/javadsl/api/src/main/java/com/lightbend/lagom/javadsl/api/transport/TransportException.java

```diff
@@ -76,21 +76,27 @@ public ExceptionMessage exceptionMessage() {
     private static final Map<TransportErrorCode, BiFunction<TransportErrorCode, ExceptionMessage, TransportException>> BY_CODE_TRANSPORT_EXCEPTIONS;
 
     static {
+        // this map keeps a more strict relation between exception message and exception instances.
+        // Some exceptions reuse the same status code on certain transports so deserialization should
+        // try to reconstruct the exception by name first and fallback to reconstructing by code.
         Map<String, BiFunction<TransportErrorCode, ExceptionMessage, TransportException>> byName = new HashMap<>();
         byName.put(DeserializationException.class.getSimpleName(), DeserializationException::new);
-        byName.put(SerializationException.class.getSimpleName(), SerializationException::new);
-        byName.put(UnsupportedMediaType.class.getSimpleName(), UnsupportedMediaType::new);
-        byName.put(NotAcceptable.class.getSimpleName(), NotAcceptable::new);
+        byName.put(BadRequest.class.getSimpleName(), BadRequest::new);
+        byName.put(Forbidden.class.getSimpleName(), Forbidden::new);
         byName.put(PolicyViolation.class.getSimpleName(), PolicyViolation::new);
         byName.put(NotFound.class.getSimpleName(), NotFound::new);
-        byName.put(Forbidden.class.getSimpleName(), Forbidden::new);
+        byName.put(NotAcceptable.class.getSimpleName(), NotAcceptable::new);
+        byName.put(PayloadTooLarge.class.getSimpleName(), PayloadTooLarge::new);
+        byName.put(SerializationException.class.getSimpleName(), SerializationException::new);
+        byName.put(UnsupportedMediaType.class.getSimpleName(), UnsupportedMediaType::new);
 
         Map<TransportErrorCode, BiFunction<TransportErrorCode, ExceptionMessage, TransportException>> byCode = new HashMap<>();
         byCode.put(DeserializationException.ERROR_CODE, DeserializationException::new);
-        byCode.put(UnsupportedMediaType.ERROR_CODE, UnsupportedMediaType::new);
-        byCode.put(NotAcceptable.ERROR_CODE, NotAcceptable::new);
-        byCode.put(PolicyViolation.ERROR_CODE, PolicyViolation::new);
         byCode.put(Forbidden.ERROR_CODE, Forbidden::new);
+        byCode.put(PolicyViolation.ERROR_CODE, PolicyViolation::new);
+        byCode.put(NotAcceptable.ERROR_CODE, NotAcceptable::new);
+        byCode.put(PayloadTooLarge.ERROR_CODE, PayloadTooLarge::new);
+        byCode.put(UnsupportedMediaType.ERROR_CODE, UnsupportedMediaType::new);
```

Comment:
Sorts the code by status code value (400, 403, 404...). With that sorting convention it's easier to detect when something is missing.

This list is shorter than the list above because the key of this map is an integer (status code) which is reused in many cases. The exceptions used for each supported integer are the most generic exception on each case. For example: `PolicyViolation` is the [general case](https://github.com/lagom/lagom/blob/6c84fc553f5ae1b18ffbe25e7f52cfbc738b1153/service/javadsl/api/src/main/java/com/lightbend/lagom/javadsl/api/transport/TransportErrorCode.java#L54-L62) for a `NotFound`.